libraries/botbuilder-azure-blobs/src/blobsStorage.ts

```diff
@@ -0,0 +1,118 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import getStream from 'get-stream';
+import pmap from 'p-map';
+import { ContainerClient, StoragePipelineOptions } from '@azure/storage-blob';
+import { Storage, StoreItems, isStoreItems } from 'botbuilder';
+import { ignoreError, isStatusCodeError } from './ignoreError';
+import { sanitizeBlobKey } from './sanitizeBlobKey';
+
+/**
+ * BlobsStorage provides a [Storage](xref:botbuilder-core.Storage) implementation backed by Azure Blob Storage
+ */
+export class BlobsStorage implements Storage {
+    private readonly _containerClient: ContainerClient;
+    private readonly _concurrency: number;
+    private _initializePromise: Promise<unknown>;
+
+    /**
+     * Constructs a BlobsStorage instance.
+     *
+     * @param connectionString Azure Blob Storage connection string
+     * @param containerName Azure Blob Storage container name
+     * @param options Azure Blob Storage [StoragePipelineOptions](xref:@azure/storage-blob.StoragePipelineOptions) options
+     */
+    constructor(connectionString: string, containerName: string, options?: StoragePipelineOptions) {
```

Comment:
Do we want to encapsulate the parameters into a single property bag? If we have to add new constructor properties then we won't get into the business of overloads. Additionally we're at three parameters, and I _think_ our rule of thumb is to stop before we get to four. Perhaps @gabog or @johnataylor has some time to weigh in on the number of parameters...

Reply:
I'm of the opinion that required parameters ought to exist separate from an options bag. It's easier to express, especially outside of Typescript (without the ability to consume an interface with both required and optional keys). I think I'll nest the `StoragePipelineOptions` under a generic `BlobsStorageOption` argument, though, which should make it easier to add other options down the line if necessary.

Reply:
Good point about optional vs required parameters, and another point about creating an options bag in TypeScript vs JavaScript... However we would up with a new overload if we had to add a required parameter to the constructor, otherwise we have required parameters outside and inside of the optional options bag

I do feel a bit odd about nesting the StoragePipelineOptions object under a generic argument since it'll be the only case that is currently nested, but otherwise it makes sense with background context.

Reply:
In general, adding required parameters _should_ be difficult because it is inherently complicated. Optimizing for our own ergonomics just pushes this complexity off to the user. If we need to add a required parameter, we should be responsible for handling the overload cases.

Reply:
After thinking about this more, should we add ContainerClient to `BlobsStorageOptions`? I'm not a fan of wrapping the Azure libraries with thin business logic, partially because we end up covering features that end users might want to utilize in their bot. Our constructor logic would have a snippet of code that looks like this:

```ts
const { containerClient, storagePipelineOptions } = options;

if (containerClient instanceof ContainerClient) {
    this._containerClient = containerClient;
} else {
    this._containerClient = new ContainerClient(connectionString, containerName, options?.storagePipelineOptions);
}
```

This logic is incomplete, because if they have a ContainerClient, the container likely already exists (so we don't need to initialize it and don't need containerName)...

@joshgummersall, @EricDahlvang what do you think?

Reply:
Not a bad idea, I'm all for it. 

The initialize promise is a no-op if the container exists (though it will more than likely make an HTTP request the first time the promise is fired off). I don't think there's much harm in leaving it in either way.